### ISSUE TITLE
[aarch64]: `cfunction: closures are not supported on this platform` breaks JuMP

### ISSUE THREAD

**Issue by staticfloat** [JuliaLang/julia#42269]
While trying to benchmark JuMP solvers on M1, we found that JuMP requires a compiler feature that is apparently not available on aarch64:

```
$ julia --project=. bench.jl 
ERROR: LoadError: cfunction: closures are not supported on this platform
Stacktrace:
  [1] _set_callback
    @ ~/.julia/dev/GLPK/src/MOI_wrapper/MOI_wrapper.jl:259 [inlined]
  [2] empty!(model::GLPK.Optimizer)
    @ GLPK ~/.julia/dev/GLPK/src/MOI_wrapper/MOI_wrapper.jl:290
  [3] GLPK.Optimizer(; want_infeasibility_certificates::Bool, method::GLPK.MethodEnum, kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ GLPK ~/.julia/dev/GLPK/src/MOI_wrapper/MOI_wrapper.jl:206
  [4] Optimizer
    @ ~/.julia/dev/GLPK/src/MOI_wrapper/MOI_wrapper.jl:171 [inlined]
  [5] _instantiate_and_check(optimizer_constructor::Type{GLPK.Optimizer})
    @ MathOptInterface ~/.julia/packages/MathOptInterface/YDdD3/src/instantiate.jl:62
  [6] instantiate(optimizer_constructor::Type; with_bridge_type::Type{Float64}, with_names::Bool)
    @ MathOptInterface ~/.julia/packages/MathOptInterface/YDdD3/src/instantiate.jl:120
  [7] set_optimizer(model::Model, optimizer_constructor::Type; bridge_constraints::Bool)
    @ JuMP ~/.julia/packages/JuMP/klrjG/src/optimizer_interface.jl:109
  [8] Model(optimizer_factory::Type; bridge_constraints::Bool, kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ JuMP ~/.julia/packages/JuMP/klrjG/src/JuMP.jl:287
  [9] Model
    @ ~/.julia/packages/JuMP/klrjG/src/JuMP.jl:286 [inlined]
 [10] example_factory_schedule(factory::Type)
    @ Main ~/mlp_bench/bench.jl:557
 [11] top-level scope
    @ ~/mlp_bench/bench.jl:605
in expression starting at /Users/sabae/mlp_bench/bench.jl:602
```

The relevant code can be found here: https://github.com/jump-dev/GLPK.jl/blob/ebc73abbf0c8352c0509932aa1ad57a69716a2e6/src/MOI_wrapper/MOI_wrapper.jl#L242-L259

Reading through the docs, it appears [this is advertised](https://docs.julialang.org/en/v1/manual/calling-c-and-fortran-code/#Closure-cfunctions), but since such a prominent package uses them internally, it seems like we might want to see if there's a workaround we can use.

**Comment by freemin7** [comment]
A list of the JuMP maintained solvers that use `@cfunctions` is available here: https://github.com/search?q=org%3Ajump-dev+cfunction%28&type=code it covers all the popular solvers written in C.

**Comment by odow** [comment]
A few of these use-cases are mission-critical, like in Ipopt. The others are for callbacks which could be optional. 

In each case, we need some way of calling from C back into Julia functions. Are there docs on how we can work-around this?

**Comment by vtjnash** [comment]
You might want to file with those packages. From a quick look, it appears that the MOI_wrapper overcomplicates their implementation to use this feature, which is both slower and less widely supported.

**Comment by odow** [comment]
> overcomplicates their implementation to use this feature

Probably. Much of the cfunction related code is Julia ~0.3 days. What's the rule for cfunction? We can use them just not with closures?

**Comment by vtjnash** [comment]
use them like 0.3, but the rules are just simplified now and syntactic. in this case, there's already a context pointer, so just use it for everything, instead of only half of things

**Comment by odow** [comment]
Okay, pretty simple fix: https://github.com/jump-dev/GLPK.jl/pull/190

Although @staticfloat you won't be able to use this branch because we're in the middle of a breaking transition throughout JuMP. However, if it's urgent, I'll back port it on to a new release.

**Comment by odow** [comment]
There's actually quite a few usages of the cfunction closures throughout the ecosystem: 
https://juliahub.com/ui/Search?q=%40cfunction%5C%28%5C%24&type=code&r=true

**Comment by vchuravy** [comment]
@staticfloat and I looked at this for a bit and we came to two conclusion:

1. Implementing the trampoline setup without LLVM support is tedious, but doable.
2. LLVM  uses X18 in it's calling convention for the nest parameter (the closure pointer), on Darwin (and Windows for that matter) that register is used as the platform register and thus we can't use it. 
3. It looks like neither GCC nor clang support the C built-in on Darwin.
